pywr-schema/src/metric_sets/mod.rs

```diff
@@ -28,11 +29,82 @@ impl OutputMetric {
     }
 }
 
+/// Aggregation function to apply over metric values.
+#[derive(serde::Deserialize, serde::Serialize, Debug, Copy, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum MetricAggFunc {
+    Sum,
+    Max,
+    Min,
+    Mean,
+}
+
+impl From<MetricAggFunc> for pywr_core::recorders::AggregationFunction {
+    fn from(value: MetricAggFunc) -> Self {
+        match value {
+            MetricAggFunc::Sum => pywr_core::recorders::AggregationFunction::Sum,
+            MetricAggFunc::Max => pywr_core::recorders::AggregationFunction::Max,
+            MetricAggFunc::Min => pywr_core::recorders::AggregationFunction::Min,
+            MetricAggFunc::Mean => pywr_core::recorders::AggregationFunction::Mean,
+        }
+    }
+}
+
+#[derive(serde::Deserialize, serde::Serialize, Debug, Copy, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum MetricAggFrequency {
+    Monthly,
+    Annual,
+}
+
+impl From<MetricAggFrequency> for pywr_core::recorders::AggregationFrequency {
+    fn from(value: MetricAggFrequency) -> Self {
+        match value {
+            MetricAggFrequency::Monthly => pywr_core::recorders::AggregationFrequency::Monthly,
+            MetricAggFrequency::Annual => pywr_core::recorders::AggregationFrequency::Annual,
+        }
+    }
+}
+
+/// A set of metrics that can be output from a model run.
+///
+/// A metric set can optionally have an aggregator, which will apply an aggregation function
+/// over the metrics in the set. If an aggregation frequency is provided then the aggregation
+/// will be performed over each period implied by that frequency. For example, if the frequency
+/// is monthly then the aggregation will be performed over each month in the model run.
 ///
+/// If the metric set has a child aggregator then the aggregation will be performed over the
+/// aggregated values of the child aggregator.
+#[derive(Deserialize, Serialize, Clone)]
+pub struct MetricAggregator {
+    /// Optional aggregation frequency.
+    freq: Option<MetricAggFrequency>,
+    /// Aggregation function to apply over metric values.
+    func: MetricAggFunc,
+    /// Optional child aggregator.
+    child: Option<Box<MetricAggregator>>,
+}
+
+impl From<MetricAggregator> for pywr_core::recorders::Aggregator {
+    fn from(value: MetricAggregator) -> Self {
+        pywr_core::recorders::Aggregator::new(
+            value.freq.map(|p| p.into()),
+            value.func.into(),
+            value.child.map(|a| (*a).into()),
+        )
+    }
+}
+
+/// A set of metrics that can be output from a model run.
+///
+/// A metric set can optionally have an aggregator, which will apply an aggregation function
+/// over metrics set. If the aggregator has a defined frequency then the aggregation will result
+/// in multiple values (i.e. per each period implied by the frequency).
```

Comment:
I think it would be good to have a doc text model here that shows how the nest aggregation works

Reply:
I wonder if this would be better as an extended section in the proposed book? I am tempted to leave it for now, but then also need to ensure we write that documentation.